lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -2,104 +2,116 @@ define((require, exports, module) ->
   LatexParsingContext = require("ace/ext/papeeria/latex_parsing_context")
   TokenIterator = require("ace/token_iterator").TokenIterator
   Range = require("ace/range").Range
+  findSurroundingBrackets = require("ace/ext/papeeria/highlighter").findSurroundingBrackets
+
+
+  class ConstrainedTokenIterator
+    constructor: (@session, @range, row, column) ->
+      @tokenIterator = new TokenIterator(@session, row, column)
+      @expired = not @range.contains(row, column)
+
+    getCurrentToken: -> if not @expired then @tokenIterator.getCurrentToken() else null
+
+    getCurrentTokenPosition: -> if not @expired then @tokenIterator.getCurrentTokenPosition() else null
+
+    stepBackward: ->
+      @tokenIterator.stepBackward()
+      curToken = @tokenIterator.getCurrentToken()
+      {row: tokenRow, column: tokenColumn} = @tokenIterator.getCurrentTokenPosition()
+      tokenRange = new Range(tokenRow, tokenColumn, tokenRow, tokenColumn + curToken.value.length)
+      if not @range.containsRange(tokenRange)
+        @tokenIterator.stepForward()
+        @expired = true
+        return null
+      else
+        return @tokenIterator.getCurrentToken()
+
+    stepForward: ->
+      @tokenIterator.stepForward()
+      curToken = @tokenIterator.getCurrentToken()
+      {row: tokenRow, column: tokenColumn} = @tokenIterator.getCurrentTokenPosition()
+      tokenRange = new Range(tokenRow, tokenColumn, tokenRow, tokenColumn + curToken.value.length)
+      if not @range.containsRange(tokenRange)
+        @tokenIterator.stepBackward()
+        @expired = true
+        return null
+      else
+        return @tokenIterator.getCurrentToken()
+
+
+  class EquationRangeHandler
+    @BEGIN_EQUATION_TOKEN_SEQUENCE: [
+      { type: "storage.type", value: "\\begin" }
+      { type: "lparen", value: "{" }
+      { type: "variable.parameter", value: "equation" }
+      { type: "rparen", value: "}" }
+    ]
+    @END_EQUATION_TOKEN_SEQUENCE: [
+      { type: "storage.type", value: "\\end" }
+      { type: "lparen", value: "{" }
+      { type: "variable.parameter", value: "equation" }
+      { type: "rparen", value: "}" }
+    ]
+
+    constructor: (@editor) ->
+
+    equalTokens: (token1, token2) ->
+      if token1? and token2?
+        return token1.type == token2.type and token1.value == token2.value
+      else return if token1? or token2? then false else true
+
+    getEquationStart: (tokenIterator) ->
```

Comment:
Yeah, it actually happened quite a lot when I was debugging it :). Ok, I'll add it.